### PR TITLE
Auth middleware

### PR DESCRIPTION
**Pull request by wyattjoh**
This PR adds a JWT authentication layer to the corald service.

`CORAL_AUTH_PUBLIC_KEY` is provided when you want authentication to be enabled and consists of a base64 encoded string of the public RSA key.

### CODE REVIEW

File: cmd/corald/midware/auth.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+var (
+	// ErrInvalidToken is returned when the token provided is not valid.
+	ErrInvalidToken = errors.New("invalid token")
+
+	// ErrInvalidClaims is returned when the claims inside a valid token are not
+	// valid.
+	ErrInvalidClaims = errors.New("invalid claims")
+)
+
+const cfgAuthPublicKey = "AUTH_PUBLIC_KEY"
```
Comment (dwhitena) [inline cmd/corald/midware/auth.go]: Comment would be nice, even for unexported stuff.

File: cmd/corald/midware/auth.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+			tokenString := c.Request.Header.Get("Authorization")
+
+			if tokenString == "" {
+				log.Error(c.SessionID, "Auth", ErrInvalidToken, "No token on request")
+				return ErrInvalidToken
+			}
+
+			token, err := jwt.ParseWithClaims(tokenString, &jwt.MapClaims{}, func(token *jwt.Token) (interface{}, error) {
+
+				// Don't forget to validate the alg is what you expect:
```
Comment (dwhitena) [inline cmd/corald/midware/auth.go]: Comment grammar.

File: cmd/corald/midware/auth.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	// valid.
+	ErrInvalidClaims = errors.New("invalid claims")
+)
+
+const cfgAuthPublicKey = "AUTH_PUBLIC_KEY"
+
+// authOff is used when authentication is turned off by not providing a public
+// key in the environment.
+func authOff(h app.Handler) app.Handler {
+	return func(c *app.Context) error {
```
Comment (ardan-bkennedy) [inline cmd/corald/midware/auth.go]: I know I am crazy about this but could we create a variable for the function literal and return that?

File: cmd/corald/midware/auth.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+// Auth handles token authentication.
+func Auth() (app.Middleware, error) {
+
+	// Load in the public key to validate the JWT tokens.
+	publicKeyBase64Str, err := cfg.String(cfgAuthPublicKey)
+	if err != nil {
+		return authOff, nil
+	}
+
+	publicKeyPEM, err := base64.StdEncoding.DecodeString(publicKeyBase64Str)
```
Comment (ardan-bkennedy) [inline cmd/corald/midware/auth.go]: Can you describe what you are doing here and why? The workflow is what is important.

File: cmd/corald/midware/auth.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+	if err != nil {
+		log.Error("startup", "Auth", err, "Can not setup Auth middleware")
+		return nil, err
+	}
+
+	publicKey, err := jwt.ParseRSAPublicKeyFromPEM(publicKeyPEM)
```
Comment (ardan-bkennedy) [inline cmd/corald/midware/auth.go]: Same as above. Comments the workflow.

File: cmd/corald/midware/auth.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+			// Ensure that the claims that are inside the token are indeed the MapClaims
+			// that we expect.
+			claims, ok := token.Claims.(*jwt.MapClaims)
+			if !ok {
+				log.Error(c.SessionID, "Auth", ErrInvalidClaims, "Claims not valid")
+				return ErrInvalidClaims
+			}
+
+			// Validate that all the parameters we expect are correct, noteably, the
+			// expiry date, and not before claims should be verified.
```
Comment (ardan-bkennedy) [inline cmd/corald/midware/auth.go]: Fix sentence.

Comment (ardan-bkennedy) [inline cmd/corald/midware/auth.go]: Wait, my mistake.

**Pull request merged**